### PR TITLE
Feature/launch darkly options

### PR DESCRIPTION
**Pull request by marcellodesales**
* Solving for #41 
* Added `README` updates for clients
* Providing the example on how to integrate

# Prop Options

screen shot 2017-10-10 at 3 37 01 pm

# LaunchDarkly clientOptions

screen shot 2017-10-10 at 3 36 43 pm

# Console Logs

screen shot 2017-10-10 at 4 32 37 pm

# Different URL Server - User

screen shot 2017-10-10 at 4 34 31 pm

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -79,6 +79,10 @@ This is the client id that is provided to you by LaunchDarkly.
 ##### `user` : `object` (required)
 See the [LaunchDarkly docs](http://docs.launchdarkly.com/docs/js-sdk-reference#section-users) for more info.
 
+##### `clientOptions` : `object` (optional)
+This is an advanced option required when pointing to an Enterprise version of launch darkly (different URLs).
```
[the remark this answers is not in the document]
Comment (marcellodesales) [reply README.md]: @jacobmoretti Thanks for reviewing it... 
* I will add the test cases for the `clientOptions` as suggested... 
* Let me know how to sign the CLA
* I will use your suggested documentation on the README.

File: src/components/LaunchDarkly.js [hunk 2]
```diff
@@ -22,7 +26,8 @@ export default function LaunchDarkly (props:Props) {
   if (clientId && user) {
     config = {
       clientId,
-      user
+      user,
+      clientOptions: (clientOptions || {})
```
[the remark this answers is not in the document]
Comment (marcellodesales) [reply src/components/LaunchDarkly.js]: Sounds good... Will remove that.

File: src/lib/utils.js [hunk 3]
```diff
@@ -10,11 +10,11 @@ export function getLocation() {
 
 let ldClient;
 let ldClientReady = false;
-export function ldClientWrapper (key, user) {
+export function ldClientWrapper (key, user, options) {
```
[the remark this answers is not in the document]
Comment (marcellodesales) [reply src/lib/utils.js]: Yeah, I can revert the tests... and update this.

**Comment by marcellodesales** [comment]
# Testing Options Failing

* The following test cases for Options fails due to the default initialization.
 * According to https://github.com/facebook/jest/issues/907, we should add the dependency `babel-jest`. I did it, but it still fails.
 * I also tried to update jest to the latest, and it still fails

```diff
diff --git a/test/lib/utils.test.js b/test/lib/utils.test.js
index 34ccd8f..1b4ffd2 100644
--- a/test/lib/utils.test.js
+++ b/test/lib/utils.test.js
@@ -10,7 +10,7 @@ describe("lib/utils", () => {
     expect(utils["getAllFeatureFlags"]).toBeDefined();
   });
 
-  describe("ldClientWrapper", () => {
+  describe("ldClientWrapper Without Options", () => {
     const key = "my key";
     const user = "my user";
 
@@ -53,6 +53,54 @@ describe("lib/utils", () => {
     });
   });
 
+  describe("ldClientWrapper with Options", () => {
+    jest.disableAutomock();
+    const key = "my key";
+    const user = "my user";
+    const options = {
+      baseUrl: "https://app.example.com",
+      streamUrl: "https://stream.example.com",
+      eventsUrl: "https://events.example.com"
+    };
+
+    launchDarklyBrowser.initialize = jest.fn().mockImplementation(() => ({
+      on: (event, callback) => {
+        setTimeout(() => {
+          callback();
+        }, 1000);
+      }
+    }));
+
+    it("proxies to ldclient-js", () => {
+      utils.ldClientWrapper(key, user, options);
+      expect(launchDarklyBrowser.initialize).toBeCalledWith(key, user, options);
+    });
+
+    it("does not instantiate ldclient-js more than once", () => {
+      utils.ldClientWrapper(key, user, options);
+      utils.ldClientWrapper(key, user, options);
+      utils.ldClientWrapper(key, user, options);
+      expect(launchDarklyBrowser.initialize).toHaveBeenCalledTimes(1);
+    });
+
+    it("onReady calls wait for ready event to fire", () => {
+      const callbackMock = jest.fn();
+      utils.ldClientWrapper(key, user, options).onReady(callbackMock);
+      utils.ldClientWrapper(key, user, options).onReady(callbackMock);
+      utils.ldClientWrapper(key, user, options).onReady(callbackMock);
+      expect(callbackMock.mock.instances.length).toBe(0);
+    });
+
+    it("executes all onReady calls after ldClient is ready", () => {
+      const callbackMock = jest.fn();
+      utils.ldClientWrapper(key, user, options).onReady(callbackMock);
+      utils.ldClientWrapper(key, user, options).onReady(callbackMock);
+      utils.ldClientWrapper(key, user, options).onReady(callbackMock);
+      jest.runAllTimers();
+      expect(callbackMock.mock.instances.length).toBe(3);
+    });
+  });
+
   describe("getAllFeatureFlags", () => {
 
   });

```

# Failures

* Could you accept the PR and work on the test cases updates?
 * I won't be able to invest more time on this after today.

screen shot 2017-10-11 at 12 05 17 pm

**Comment by jacobmoretti** [comment]
[clabot:check]

**Comment by jacobmoretti** [comment]
@marcellodesales I can finish up the tests. Having a few problems with setting up our CLA Bot. When its available, I'll ping you again, hopefully you can sign it then?

**Comment by marcellodesales** [comment]
@jacobmoretti I will definitely sign it! Thanks for taking care of the tests... Spent the morning trying to figure out the `babel-jest` settings to by-pass the default param method without success...

**Comment by jacobmoretti** [comment]
[clabot:check]

**Comment by jacobmoretti** [comment]
[clabot:check]

**Comment by jacobmoretti** [comment]
[clabot:check]

**Comment by jacobmoretti** [comment]
[clabot:check]

**Comment by marcellodesales** [comment]
[clabot:check]

**Comment by marcellodesales** [comment]
@jacobmoretti I have signed the CLA and instructions show to send a `[clabot:check]`, which I did... Not sure what the next steps are.

**Comment by jacobmoretti** [comment]
@marcellodesales Thanks for signing and contributing! I'll take it from here and hopefully have this good to go by the end of the week.

**Comment by marcellodesales** [comment]
@jacobmoretti Thanks a lot! Let me know when it's released as we have 2 teams at least waiting on this patch! 👍 

I look forward to seeing your solution about the test cases. I think you need to transpile the source during tests before you run them.

**Comment by jacobmoretti** [comment]
Merging this into `develop` as is and will add some tests in a subsequent PR. This brings to light that I should probably add a CONTRIBUTING doc and an actual LICENSE.

**Comment by marcellodesales** [comment]
@jacobmoretti yeah, I selected the default branch 👍 So, if you need anything else let me know!

**Pull request merged**